crates/weaver_resolver/src/registry.rs

```diff
@@ -203,6 +238,39 @@ fn check_group_any_of_constraints(
     Ok(())
 }
 
+/// Generic function to check for duplicate keys in the given registry.
+///
+/// A key can be a group ID, a metric name, an event name, or any other key that is used
+/// to identify a group.
+fn check_uniqueness<K, KF, EF>(
+    registry: &Registry,
+    errors: &mut Vec<Error>,
+    key_fn: KF,
+    error_fn: EF,
+) where
+    K: Eq + Display + Hash,
+    KF: Fn(&Group) -> Option<K>,
+    EF: Fn(String, Vec<String>) -> Error,
+{
+    let mut keys: HashMap<K, Vec<String>> = HashMap::new();
+
+    for group in registry.groups.iter() {
+        if let Some(key) = key_fn(group) {
+            let provenances = keys.entry(key).or_default();
+            provenances.push(group.provenance().to_owned());
+        }
+    }
+
+    for (key, provenances) in keys {
+        if provenances.len() > 1 {
+            // Deduplicate the provenances.
+            let provenances: HashSet<String> = provenances.into_iter().collect();
```

Comment:
Nit: You can use itertools to `.unique` this when collector instead of forcing a hashset.

I think performance wise it's the same, just easier to read.